server/monitor-types/snmp.js

```diff
@@ -0,0 +1,102 @@
+const { MonitorType } = require("./monitor-type");
+const { UP, DOWN, log, getKey } = require("../../src/util");
+const snmp = require("net-snmp");
+
+class SNMPMonitorType extends MonitorType {
+    name = "snmp";
+
+    /**
+     * @inheritdoc
+     */
+    async check(monitor, heartbeat, _server) {
+
+        const options = {
+            port: monitor.port || "161",
+            retries: monitor.maxretries,
+            timeout: monitor.timeout * 1000,
+
+            // Resolve the net-snmp version identifier from monitor.snmpVersion using the `getKey` method (see utils.ts).
+            // If the specified version exists, use it; otherwise, default to SNMP version 2c.
+            version: getKey(snmp.Version, monitor.snmpVersion) || snmp.Version2c,
+        };
+
+        try {
+            const session = snmp.createSession(monitor.hostname, monitor.snmpCommunityString, options);
+
+            // Handle errors during session creation
+            session.on("error", (error) => {
+                heartbeat.status = DOWN;
+                heartbeat.msg = `SNMP: Error creating SNMP session: ${error.message}`;
+                log.debug("monitor", heartbeat.msg);
+            });
+
+            const varbinds = await new Promise((resolve, reject) => {
+                session.get([ monitor.snmpOid ], (error, varbinds) => {
+                    if (error) {
+                        reject(error);
+                    } else {
+                        log.debug("monitor", `SNMP: Received varbinds (Type=${getKey(snmp.ObjectType, varbinds[0].type)}): ${varbinds[0].value}`);
+                        resolve(varbinds);
+                    }
+                });
+            });
+
+            // Verify if any varbinds were returned from the SNMP session or if the varbind type indicates a non-existent instance.
+            // The `getKey` method retrieves the key associated with the varbind type from the snmp.ObjectType object.
+            if (varbinds.length === 0 || getKey(snmp.ObjectType, varbinds[0].type) === "NoSuchInstance") {
+                throw new Error(`No varbinds returned from SNMP session (OID: ${monitor.snmpOid})`);
```

Comment:
In what case can `NoSuchInstance` occur? Is the reported error messsage accurate for this case?

Reply:
Yes, the `NoSuchInstance` type is typically returned by the net-snmp library when the SNMP query is successful, but the OID queried returned no results. So my placement of this check and associated error is correct.

The condition inmy code is effectively checking for two scenarios:

If `varbinds.length === 0`, it means that no varbinds were returned from the SNMP session, indicating that there are no results for the queried OID.
If `getKey(snmp.ObjectType, varbinds[0].type) === "NoSuchInstance"`, it implies that the varbind type of the first varbind returned by the SNMP session is `NoSuchInstance`. This typically indicates that the instance being queried does not exist.

Reply:
`no results for the queried OID` and `typically indicates that the instance being queried does not exist.` seem like different cases to me.
=> please split them and provide an error message to allow devs to debug this ^^

Reply:
Good point. Resolved with e9b52eb0e7f2133bd273a365d1ee061eaeeb6d4d